pep-0561.rst

```diff
@@ -0,0 +1,178 @@
+PEP: 561 
+Title: Distributing and Packaging Type Information
+Author: Ethan Smith <ethan@ethanhs.me>
+Status: Draft
+Type: Standards Track
+Content-Type: text/x-rst
+Created: 09-Sep-2017
+Python-Version: 3.7
+Post-History: 
+
+
+Abstract
+========
+
+PEP 484 introduced type hints to Python, with goals of making typing
+gradual and easy to adopt. Currently, typing information must be distributed 
+manually. This PEP provides a standardized means to package and distribute
+type information and an ordering for type checkers to resolve modules and 
+collect this information for type checking using existing packaging
+architecture.
+
+
+Rationale
+=========
+
+PEP 484 has a brief section on distributing typing information. In this
+section [1]_ the PEP recommends using ``shared/typehints/pythonX.Y/`` for
+shipping stub files. However, manually adding a path to stub files for each
+third party library does not scale. The simplest approach people have taken
+is to add ``site-packages`` to their ``PYTHONPATH``, but this causes type
+checkers to fail on packages that are highly dynamic (e.g. sqlalchemy 
+and Django).
+
+Furthermore, package authors are wishing to distribute code that has
+inline type information, and there currently is no standard method to
+distribute packages with inline type annotations or syntax that can
+simultaneously be used at runtime and in type checking.
+
+
+Specification
+=============
+
+There are several motivations and methods of supporting typing in a package.                                                 This PEP recognizes three (3) types of packages that may be created:
+
+1. The package maintainer would like to add type information inline.
+
+2. The package maintainer would like to add type information via stubs.
+
+3. A third party would like to share stub files for a package, but the
+   maintainer does not want to include them in the source of the package.
+   
+This PEP aims to support these scenarios and make them simple to add to packaging and deploy.
+
+The two major parts of this specification are the packaging specifications
+and the resolution order for resolving module type information. This spec
+is meant to replace the ``shared/typehints/pythonX.Y/`` spec of PEP 484 [1]_. 
+
+Packaging Type Information
+--------------------------
+
+Packages must opt into supporting typing. This will be done though a distutils
+extension [2]_, providing a ``typed`` keyword argument to the distutils
+``setup()`` command. The argument value will depend on the kind of type
+information the package provides. The distutils extension will be added to the
+``typing`` package. Therefore a package maintainer may write
+
+::
+
+    setup(
+        ...
+        setup_requires=["typing"],
+        typed="inline",
+        ...
+    )
+
+Stub Only Packages
```

Comment:
I would move this section below the next one so that the order will match the order of three numbered options above.